<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Datasets" data-toc-modified-id="Datasets-1">Datasets</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Process-shared_articles.csv" data-toc-modified-id="Process-shared_articles.csv-1.0.1">Process <code>shared_articles.csv</code></a></span></li><li><span><a href="#Process-users_interactions.csv" data-toc-modified-id="Process-users_interactions.csv-1.0.2">Process <code>users_interactions.csv</code></a></span></li></ul></li></ul></li><li><span><a href="#Data-Preparation" data-toc-modified-id="Data-Preparation-2">Data Preparation</a></span><ul class="toc-item"><li><span><a href="#Build-Profile-for-Each-User" data-toc-modified-id="Build-Profile-for-Each-User-2.1">Build Profile for Each User</a></span></li></ul></li><li><span><a href="#Develop-a-Content-Based-Recommendation-System" data-toc-modified-id="Develop-a-Content-Based-Recommendation-System-3">Develop a Content-Based Recommendation System</a></span></li><li><span><a href="#Matrix-Factorization" data-toc-modified-id="Matrix-Factorization-4">Matrix Factorization</a></span></li><li><span><a href="#Hybrid-Recommender" data-toc-modified-id="Hybrid-Recommender-5">Hybrid Recommender</a></span></li><li><span><a href="#Comparing-the-methods" data-toc-modified-id="Comparing-the-methods-6">Comparing the methods</a></span></li></ul></div>

In [3]:
import numpy as np
import scipy
import pandas as pd
import math
import random
import sklearn
# from nltk.corpus import stopwords
from scipy.sparse import csr_matrix
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity
from scipy.sparse.linalg import svds
from sklearn.preprocessing import MinMaxScaler
import matplotlib.pyplot as plt
import os

In [207]:
DS_UTILS_PATH=os.environ.get('DS_UTILS')

In [208]:
%run {DS_UTILS_PATH}

In [209]:
import warnings
warnings.filterwarnings('ignore')

# Recommender Systems

- [Lecture 41 — Overview of Recommender Systems | Stanford University](https://www.youtube.com/watch?v=1JRrCEgiyHM&list=PLLssT5z_DsK9JDLcT8T62VtzwyW9LNepV&index=41&t=84s)

This notebook is a practical introduction to the main [Recommender System](https://en.wikipedia.org/wiki/Recommender_system) (RecSys) techniques. The objective of a RecSys is to recommend relevant items for users, based on their preference. Preference and relevance are subjective, and they are generally inferred by items users have consumed previously.  
The main families of methods for RecSys are:  
- [**Collaborative Filtering**](https://en.wikipedia.org/wiki/Collaborative_filtering): This method makes automatic predictions (filtering) about the interests of a user by collecting preferences or taste information from many users (collaborating). The underlying assumption of the collaborative filtering approach is that if a person A has the same opinion as a person B on a set of items, A is more likely to have B's opinion for a given item than that of a randomly chosen person.   
- [**Content-Based Filtering**](http://recommender-systems.org/content-based-filtering/): This method uses only information about the description and attributes of the items users has previously consumed to model user's preferences. In other words, these algorithms try to recommend items that are similar to those that a user liked in the past (or is examining in the present). In particular, various candidate items are compared with items previously rated by the user and the best-matching items are recommended.  
- **Hybrid methods**:  Recent research has demonstrated that a hybrid approach, combining collaborative filtering and content-based filtering could be more effective than pure approaches in some cases. These methods can also be used to overcome some of the common problems in recommender systems such as cold start and the sparsity problem.

## Datasets

In this notebook, we use a dataset we've shared on Kaggle Datasets: [Articles Sharing and Reading from CI&T Deskdrop](https://www.kaggle.com/gspmoreira/articles-sharing-reading-from-cit-deskdrop).  
We will demonstrate how to implement **Collaborative Filtering**, **Content-Based Filtering** and **Hybrid methods** in Python, for the task of providing personalized recommendations to the users.

In this section, we load the [Deskdrop dataset](https://www.kaggle.com/gspmoreira/articles-sharing-reading-from-cit-deskdrop), which contains a real sample of 12 months logs (Mar. 2016 - Feb. 2017) from CI&T's Internal Communication platform (DeskDrop). It contains about `73k logged users interactions` on more than `3k public articles` shared in the platform.
It is composed of two CSV files:  
- **shared_articles.csv**
- **users_interactions.csv**

Take a look in this kernels for a better picture of the dataset: 
- Deskdrop datasets EDA 
- DeskDrop Articles Topic Modeling

#### Process `shared_articles.csv`

Contains information about the articles shared in the platform. Each article has its sharing date (timestamp), the original url, title, content in plain text, the article' lang (Portuguese: pt or English: en) and information about the user who shared the article (author).

There are two possible event types at a given timestamp: 
- CONTENT SHARED: The article was shared in the platform and is available for users. 
- CONTENT REMOVED: The article was removed from the platform and not available for further recommendation.

For the sake of simplicity, we only consider here the "CONTENT SHARED" event type, assuming (naively) that all articles were available during the whole one year period. For a more precise evaluation (and higher accuracy), only articles that were available at a given time should be recommended, but we let this exercice for you.

In [4]:
PATH=os.environ['DATA']

In [5]:
articles_df = pd.read_csv(PATH+'/shared_articles.csv')
articles_df = articles_df[articles_df['eventType'] == 'CONTENT SHARED']
articles_df.head(3)

timestamp       eventType            contentId       authorPersonId  \
1  1459193988  CONTENT SHARED -4110354420726924665  4340306774493623681   
2  1459194146  CONTENT SHARED -7292285110016212249  4340306774493623681   
3  1459194474  CONTENT SHARED -6151852268067518688  3891637997717104548   

       authorSessionId authorUserAgent authorRegion authorCountry contentType  \
1  8940341205206233829             NaN          NaN           NaN        HTML   
2  8940341205206233829             NaN          NaN           NaN        HTML   
3 -1457532940883382585             NaN          NaN           NaN        HTML   

                                                 url  \
1  http://www.nytimes.com/2016/03/28/business/dea...   
2  http://cointelegraph.com/news/bitcoin-future-w...   
3  https://cloudplatform.googleblog.com/2016/03/G...   

                                               title  \
1  Ethereum, a Virtual Currency, Enables Transact...   
2  Bitcoin Future: When GBPcoin of Branson Wins O...   
3                       Google Data Center 360° Tour   

                                                text lang  
1  All of this work is still very early. The firs...   en  
2  The alarm clock wakes me at 8:00 with stream o...   en  
3  We're excited to share the Google Data Center ...   en

In [212]:
articles_df.shape

(3047, 13)

In [213]:
# articles_df.text.iloc[0]

#### Process `users_interactions.csv`

Contains logs of user interactions on shared articles. It can be joined to **articles_shared.csv** by **contentId** column.

The eventType values are:  
- **VIEW**: The user has opened the article. 
- **LIKE**: The user has liked the article. 
- **COMMENT CREATED**: The user created a comment in the article. 
- **FOLLOW**: The user chose to be notified on any new comment in the article. 
- **BOOKMARK**: The user has bookmarked the article for easy return in the future.

In [214]:
interactions_df = pd.read_csv(PATH+'/users_interactions.csv')
interactions_df.head(3)

timestamp eventType            contentId             personId  \
0  1465413032      VIEW -3499919498720038879 -8845298781299428018   
1  1465412560      VIEW  8890720798209849691 -1032019229384696495   
2  1465416190      VIEW   310515487419366995 -1130272294246983140   

             sessionId                                          userAgent  \
0  1264196770339959068                                                NaN   
1  3621737643587579081  Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_2...   
2  2631864456530402479                                                NaN   

  userRegion userCountry  
0        NaN         NaN  
1         NY          US  
2        NaN         NaN

In [215]:
interactions_df.shape

(72312, 8)

In [216]:
interactions_df.eventType.isnull().sum()

0

As there are different interactions types, we associate them with a weight or strength, assuming that, for example, a comment in an article indicates a higher interest of the user on the item than a like, or than a simple view.

In [217]:
event_type_strength = {
   'VIEW': 1.0,
   'LIKE': 2.0, 
   'BOOKMARK': 2.5, 
   'FOLLOW': 3.0,
   'COMMENT CREATED': 4.0,
}

In [218]:
interactions_df['eventStrength'] = interactions_df['eventType'].apply(lambda x: event_type_strength[x])

In [219]:
interactions_df.dtypes

timestamp          int64
eventType         object
contentId          int64
personId           int64
sessionId          int64
userAgent         object
userRegion        object
userCountry       object
eventStrength    float64
dtype: object

In [220]:
interactions_df.tail(3)

timestamp eventType            contentId             personId  \
72309  1485190072      VIEW -1999468346928419252 -9196668942822132778   
72310  1485190434      VIEW -6590819806697898649 -9016528795238256703   
72311  1485187988      VIEW  5669290109546991426  5713241217519616260   

                 sessionId                                          userAgent  \
72309 -8300596454915870873  Mozilla/5.0 (Windows NT 10.0; Win64; x64) Appl...   
72310  8614469745607949425  Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_4...   
72311  6704552481483665372  Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebK...   

      userRegion userCountry  eventStrength  
72309         SP          BR            1.0  
72310         MG          BR            1.0  
72311         MG          BR            1.0

In [221]:
interactions_df.shape

(72312, 9)

In [222]:
interactions_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 72312 entries, 0 to 72311
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   timestamp      72312 non-null  int64  
 1   eventType      72312 non-null  object 
 2   contentId      72312 non-null  int64  
 3   personId       72312 non-null  int64  
 4   sessionId      72312 non-null  int64  
 5   userAgent      56918 non-null  object 
 6   userRegion     56907 non-null  object 
 7   userCountry    56918 non-null  object 
 8   eventStrength  72312 non-null  float64
dtypes: float64(1), int64(4), object(4)
memory usage: 5.0+ MB


Recommender systems have a problem known as ***user cold-start***, in which is hard do provide personalized recommendations for users with none or a very few number of consumed items, due to the lack of information to model their preferences.  
For this reason, we are keeping in the dataset only users with at leas 5 interactions.

In [223]:
gruped_srs = interactions_df.groupby(['personId', 'contentId']).size(); gruped_srs # type(srs) = pandas.core.series.Series

personId              contentId           
-9223121837663643404  -8949113594875411859    1
                      -8377626164558006982    1
                      -8208801367848627943    1
                      -8187220755213888616    1
                      -7423191370472335463    8
                                             ..
 9210530975708218054   8477804012624580461    4
                       8526042588044002101    1
                       8856169137131817223    1
                       8869347744613364434    1
                       9209886322932807692    1
Length: 40710, dtype: int64

In [224]:
gruped_srs.loc[:].size

40710

In [225]:
gruped_srs[-9212075797126931087]

contentId
-1995591062742965408    2
-969155230116728853     2
 1796618361228441836    2
 5162508842577612756    1
 6852597772196653540    2
dtype: int64

In [226]:
gruped_srs.loc[-9223121837663643404].size

43

In [227]:
interactions_df.groupby(['personId', 'contentId']).size().groupby('personId').size()

personId
-9223121837663643404    43
-9212075797126931087     5
-9207251133131336884     7
-9199575329909162940    11
-9196668942822132778     7
                        ..
 9165571805999894845     5
 9187866633451383747    17
 9191849144618614467    12
 9199170757466086545    14
 9210530975708218054    63
Length: 1895, dtype: int64

In [228]:
users_interactions_count_srs = interactions_df.groupby(['personId', 'contentId']).size().groupby('personId').size()
users_interactions_count_srs

personId
-9223121837663643404    43
-9212075797126931087     5
-9207251133131336884     7
-9199575329909162940    11
-9196668942822132778     7
                        ..
 9165571805999894845     5
 9187866633451383747    17
 9191849144618614467    12
 9199170757466086545    14
 9210530975708218054    63
Length: 1895, dtype: int64

In [229]:
type(users_interactions_count_srs), users_interactions_count_srs.size

(pandas.core.series.Series, 1895)

In [230]:
users_interactions_count_srs[-9212075797126931087]

5

In [231]:
type(users_interactions_count_srs[users_interactions_count_srs >= 5])

pandas.core.series.Series

In [232]:
users_interactions_count_srs[users_interactions_count_srs >= 5].reset_index() # DataFrame Object

personId   0
0    -9223121837663643404  43
1    -9212075797126931087   5
2    -9207251133131336884   7
3    -9199575329909162940  11
4    -9196668942822132778   7
...                   ...  ..
1135  9165571805999894845   5
1136  9187866633451383747  17
1137  9191849144618614467  12
1138  9199170757466086545  14
1139  9210530975708218054  63

[1140 rows x 2 columns]

In [233]:
# Select the users of five or more interactions
selected_users = users_interactions_count_srs[users_interactions_count_srs >= 5].reset_index()[['personId']]

In [234]:
selected_users

personId
0    -9223121837663643404
1    -9212075797126931087
2    -9207251133131336884
3    -9199575329909162940
4    -9196668942822132778
...                   ...
1135  9165571805999894845
1136  9187866633451383747
1137  9191849144618614467
1138  9199170757466086545
1139  9210530975708218054

[1140 rows x 1 columns]

In [235]:
selected_users.shape

(1140, 1)

In [236]:
# Users who interected with atleast 5 contents.
selected_users

personId
0    -9223121837663643404
1    -9212075797126931087
2    -9207251133131336884
3    -9199575329909162940
4    -9196668942822132778
...                   ...
1135  9165571805999894845
1136  9187866633451383747
1137  9191849144618614467
1138  9199170757466086545
1139  9210530975708218054

[1140 rows x 1 columns]

In [237]:
interactions_df.shape

(72312, 9)

In [238]:
selected_interactions = interactions_df.merge(selected_users, how = 'right',)

In [239]:
selected_interactions.shape

(69868, 9)

In Deskdrop, users are allowed to view an article many times, and interact with them in different ways (eg. like or comment). Thus, to model the user interest on a given article, we aggregate all the interactions the user has performed in an item by a weighted sum of interaction type strength and apply a log transformation to smooth the distribution.

In [240]:
selected_interactions.groupby(['personId', 'contentId'])['eventStrength'].sum()

personId              contentId           
-9223121837663643404  -8949113594875411859    1.0
                      -8377626164558006982    1.0
                      -8208801367848627943    1.0
                      -8187220755213888616    1.0
                      -7423191370472335463    8.0
                                             ... 
 9210530975708218054   8477804012624580461    8.5
                       8526042588044002101    1.0
                       8856169137131817223    1.0
                       8869347744613364434    1.0
                       9209886322932807692    1.0
Name: eventStrength, Length: 39106, dtype: float64

In [241]:
def smooth_user_preference(x):
    return math.log(1+x, 2)

In [242]:
ratings_srs = selected_interactions.groupby(['personId', 'contentId'])['eventStrength'].sum()

In [243]:
ratings_df = ratings_srs.apply(smooth_user_preference).reset_index()

In [244]:
ratings_df.shape

(39106, 3)

In [245]:
ratings_df.head(10)

personId            contentId  eventStrength
0 -9223121837663643404 -8949113594875411859       1.000000
1 -9223121837663643404 -8377626164558006982       1.000000
2 -9223121837663643404 -8208801367848627943       1.000000
3 -9223121837663643404 -8187220755213888616       1.000000
4 -9223121837663643404 -7423191370472335463       3.169925
5 -9223121837663643404 -7331393944609614247       1.000000
6 -9223121837663643404 -6872546942144599345       1.000000
7 -9223121837663643404 -6728844082024523434       1.000000
8 -9223121837663643404 -6590819806697898649       1.000000
9 -9223121837663643404 -6558712014192834002       1.584963

In [246]:
interactions_full_df = ratings_df

# Evaluation

Evaluation is important for machine learning projects, because it allows to compare objectivelly different algorithms and hyperparameter choices for models.  
One key aspect of evaluation is to ensure that the trained model generalizes for data it was not trained on, using **Cross-validation** techniques. We are using here a simple cross-validation approach named **holdout**, in which a random data sample (20% in this case) are kept aside in the training process, and exclusively used for evaluation. All evaluation metrics reported here are computed using the **test set**.

Ps. A more robust evaluation approach could be to split train and test sets by a reference date, where the train set is composed by all interactions before that date, and the test set are interactions after that date. For the sake of simplicity, we chose the first random approach for this notebook, but you may want to try the second approach to better simulate how the recsys would perform in production predicting "future" users interactions.

In [247]:
ratings_trn_df, ratings_test_df = train_test_split(ratings_df,
                                   stratify=interactions_full_df['personId'], 
                                   test_size=0.20,
                                   random_state=42)

In Recommender Systems, there are a set metrics commonly used for evaluation. We chose to work with **Top-N accuracy metrics**, which evaluates the accuracy of the top recommendations provided to a user, comparing to the items the user has actually interacted in test set.  
This evaluation method works as follows:

* For each user
    * For each item the user has interacted in test set
        * Sample 100 other items the user has never interacted.   
        Ps. Here we naively assume those non interacted items are not relevant to the user, which might not be true, as the user may simply not be aware of those not interacted items. But let's keep this assumption.
        * Ask the recommender model to produce a ranked list of recommended items, from a set composed one interacted item and the 100 non-interacted ("non-relevant!) items
        * Compute the Top-N accuracy metrics for this user and interacted item from the recommendations ranked list
* Aggregate the global Top-N accuracy metrics

The Top-N accuracy metric choosen was **Recall@N** which evaluates whether the interacted item is among the top N items (hit) in the ranked list of 101 recommendations for a user.  
Ps. Other popular ranking metrics are **NDCG@N** and **MAP@N**, whose score calculation takes into account the position of the relevant item in the ranked list (max. value if relevant item is in the first position). You can find a reference to implement this metrics in this [post](http://fastml.com/evaluating-recommender-systems/).

In [248]:
# Indexing by personId to speed up the searches during evaluation
ratings_df_idx = ratings_df.set_index('personId')
ratings_trn = ratings_trn_df.set_index('personId')
ratings_test = ratings_test_df.set_index('personId')

In [249]:
ratings_df_idx.head()

contentId  eventStrength
personId                                                
-9223121837663643404 -8949113594875411859       1.000000
-9223121837663643404 -8377626164558006982       1.000000
-9223121837663643404 -8208801367848627943       1.000000
-9223121837663643404 -8187220755213888616       1.000000
-9223121837663643404 -7423191370472335463       3.169925

In [250]:
ratings_trn.shape, ratings_test.shape

((31284, 2), (7822, 2))

In [251]:
ratings_trn.loc[-9223121837663643404][0: 5]

contentId  eventStrength
personId                                                
-9223121837663643404  -730957269757756529            1.0
-9223121837663643404 -8377626164558006982            1.0
-9223121837663643404  5211673327552264703            1.0
-9223121837663643404   921770761777842242            1.0
-9223121837663643404 -3750879736572068916            1.0

In [252]:
def get_items_interacted(person_id, ratings_df):
    # Get the user's data and merge in the content information.
    interacted_items = ratings_df.loc[person_id]['contentId']
    return set(interacted_items if type(interacted_items) == pd.Series else [interacted_items])

In [253]:
# Top-N accuracy metrics consts
NUM_NON_INTERACTED_ITEMS = 100

class ModelEvaluator:

    def get_not_interacted_items_sample(self, person_id, sample_size, seed=42):
        interacted_items = get_items_interacted(person_id, ratings_df_idx)
        all_items = set(articles_df['contentId'])
        non_interacted_items = all_items - interacted_items

        random.seed(seed)
        non_interacted_items = random.sample(non_interacted_items, sample_size)
        return set(non_interacted_items)

    def _verify_hit_top_n(self, item_id, recommended_items, topn):
            try:
                index = next(i for i, c in enumerate(recommended_items) if c == item_id)
            except:
                index = -1
            hit = int(index in range(0, topn))
            return hit, index

    def evaluate_model_for_user(self, model, person_id):
        # Getting the items in test set
        person_ratings_test = ratings_test.loc[person_id]
        if type(person_ratings_test['contentId']) == pd.Series:
            person_items = set(person_ratings_test['contentId'])
        else:
            person_items = set([int(person_ratings_test['contentId'])])  
        
        person_items_count = len(person_items) 

        # Getting a ranked recommendation list from a model for a given user
        person_recs_df = model.recommend_items(
            person_id, items_to_ignore=get_items_interacted(person_id, 
            ratings_trn), topn=10_000_000_000
        )

        hits_at_5_count = 0
        hits_at_10_count = 0
        
        # For each item the user has interacted in test set
        for item_id in person_items:
            # Getting a random sample (100) items the user has not interacted 
            # (to represent items that are assumed to be no relevant to the user)
            non_interacted_items_sample = self.get_not_interacted_items_sample(
                person_id, sample_size=NUM_NON_INTERACTED_ITEMS, seed=item_id%(2**32)
            )

            # Combining the current interacted item with the 100 random items
            items_to_filter_recs = non_interacted_items_sample.union(set([item_id]))

            # Filtering only recommendations that are either the interacted item or from a random sample of 100 non-interacted items
            valid_recs_df = person_recs_df[person_recs_df['contentId'].isin(items_to_filter_recs)]                    
            valid_recs = valid_recs_df['contentId'].values
            # Verifying if the current interacted item is among the Top-N recommended items
            hit_at_5, index_at_5 = self._verify_hit_top_n(item_id, valid_recs, 5)
            hits_at_5_count += hit_at_5
            hit_at_10, index_at_10 = self._verify_hit_top_n(item_id, valid_recs, 10)
            hits_at_10_count += hit_at_10

        # Recall is the rate of the interacted items that are ranked among the Top-N recommended items, 
        # when mixed with a set of non-relevant items
        recall_at_5 = hits_at_5_count / float(person_items_count)
        recall_at_10 = hits_at_10_count / float(person_items_count)

        person_metrics = {'hits@5_count'    : hits_at_5_count, 
                          'hits@10_count'   : hits_at_10_count, 
                          'interacted_count': person_items_count,
                          'recall@5'        : recall_at_5,
                          'recall@10'       : recall_at_10}
        return person_metrics

    def evaluate_model(self, model):
        # print('Running evaluation for users')
        people_metrics = []
        for idx, person_id in enumerate(list(ratings_test.index.unique().values)):
            #if idx % 100 == 0 and idx > 0:
            #    print('%d users processed' % idx)
            person_metrics = self.evaluate_model_for_user(model, person_id)  
            person_metrics['_person_id'] = person_id
            people_metrics.append(person_metrics)
        print('%d users processed' % idx)

        detailed_results_df = pd.DataFrame(people_metrics) \
                            .sort_values('interacted_count', ascending=False)
        
        global_recall_at_5 = detailed_results_df['hits@5_count'].sum() / float(detailed_results_df['interacted_count'].sum())
        global_recall_at_10 = detailed_results_df['hits@10_count'].sum() / float(detailed_results_df['interacted_count'].sum())
        
        global_metrics = {'modelName': model.get_model_name(),
                          'recall@5': global_recall_at_5,
                          'recall@10': global_recall_at_10}    
        return global_metrics, detailed_results_df  

In [254]:
model_evaluator = ModelEvaluator()  

# Popularity model

A common (and usually hard-to-beat) baseline approach is the Popularity model. This model is not actually personalized - it simply recommends to a user the most popular items that the user has not previously consumed. As the popularity accounts for the "wisdom of the crowds", it usually provides good recommendations, generally interesting for most people.   
Ps. The main objective of a recommender system is to leverage the long-tail items to the users with very specific interests, which goes far beyond this simple technique.

In [255]:
interactions_full_df.groupby('contentId')['eventStrength'].sum()

contentId
-9222795471790223670    15.089906
-9216926795620865886    13.754888
-9194572880052200111    24.767202
-9192549002213406534    52.339850
-9190737901804729417     6.643856
                          ...    
 9213260650272029784     8.584963
 9215261273565326920    22.674053
 9217155070834564627    11.044394
 9220445660318725468    38.884171
 9222265156747237864     5.000000
Name: eventStrength, Length: 2984, dtype: float64

In [256]:
# Computes the most popular items
item_popularity_df = interactions_full_df.groupby('contentId')['eventStrength'].sum().sort_values(ascending=False).reset_index()
item_popularity_df.head(10)

contentId  eventStrength
0 -4029704725707465084     307.733799
1 -6783772548752091658     233.762157
2  -133139342397538859     228.024567
3 -8208801367848627943     197.107608
4 -6843047699859121724     193.825208
5  8224860111193157980     189.044680
6 -2358756719610361882     183.110951
7  2581138407738454418     180.282876
8  7507067965574797372     179.094002
9  1469580151036142903     170.548969

In [257]:
class PopularityRecommender:
    
    MODEL_NAME = 'Popularity'
    
    def __init__(self, popularity_df, items_df=None):
        self.popularity_df = popularity_df
        self.items_df = items_df
        
    def get_model_name(self):
        return self.MODEL_NAME
        
    def recommend_items(self, user_id, items_to_ignore=[], topn=10, verbose=False):
        # Recommend the more popular items that the user hasn't seen yet.
        recommendations_df = self.popularity_df[~self.popularity_df['contentId'].isin(items_to_ignore)] \
                               .sort_values('eventStrength', ascending = False) \
                               .head(topn)

        if verbose:
            if self.items_df is None:
                raise Exception('"items_df" is required in verbose mode')

            recommendations_df = recommendations_df.merge(self.items_df, how = 'left', 
                                                          left_on = 'contentId', 
                                                          right_on = 'contentId')[['eventStrength', 'contentId', 'title', 'url', 'lang']]


        return recommendations_df

In [258]:
popularity_model = PopularityRecommender(item_popularity_df, articles_df)

Here we perform the evaluation of the Popularity model, according to the method described above.  
It achieved the **Recall@5** of **0.2417**, which means that about **24%** of interacted items in test set were ranked by Popularity model among the top-5 items (from lists with 100 random items). And **Recall@10** was even higher (**37%**), as expected.  
It might be surprising to you that usually Popularity models could perform so well!

In [259]:
print('Evaluating Popularity recommendation model...')
pop_global_metrics, pop_detailed_results_df = model_evaluator.evaluate_model(popularity_model)

Evaluating Popularity recommendation model...
1139 users processed


In [260]:
print('\nGlobal metrics:\n%s' % pop_global_metrics)
pop_detailed_results_df.head(10)


Global metrics:
{'modelName': 'Popularity', 'recall@5': 0.2418818716440808, 'recall@10': 0.3725389925850166}


hits@5_count  hits@10_count  interacted_count  recall@5  recall@10  \
76             28             50               192  0.145833   0.260417   
17             12             25               134  0.089552   0.186567   
16             13             23               130  0.100000   0.176923   
10              5              9               117  0.042735   0.076923   
82             26             40                88  0.295455   0.454545   
161            12             18                80  0.150000   0.225000   
65             20             34                73  0.273973   0.465753   
81             17             23                69  0.246377   0.333333   
106            14             18                69  0.202899   0.260870   
52             21             28                68  0.308824   0.411765   

              _person_id  
76   3609194402293569455  
17  -2626634673110551643  
16  -1032019229384696495  
10  -1443636648652872475  
82  -2979881261169775358  
161 -3596626804281480007  
65   1116121227607581999  
81    692689608292948411  
106 -9016528795238256703  
52   3636910968448833585

# Content-Based Filtering model

- [Lecture 42 — Content Based Recommendations | Stanford University](https://www.youtube.com/watch?v=2uxXPzm-7FY&list=PLLssT5z_DsK9JDLcT8T62VtzwyW9LNepV&index=42)

Content-based filtering approaches leverage description or attributes from items the user has interacted to recommend similar items. It depends only on the user previous choices, making this method robust to avoid the *cold-start* problem.
For textual items, like articles, news and books, it is simple to use the raw text to build item profiles and user profiles.  
Here we are using a very popular technique in information retrieval (search engines) named [TF-IDF](https://en.wikipedia.org/wiki/Tf%E2%80%93idf). This technique converts unstructured text into a vector structure, where each word is represented by a position in the vector, and the value measures how relevant a given word is for an article. As all items will be represented in the same [Vector Space Model](https://en.wikipedia.org/wiki/Vector_space_model), it is to compute similarity between articles.  
See this [presentation](https://www.slideshare.net/gabrielspmoreira/discovering-users-topics-of-interest-in-recommender-systems-tdc-sp-2016) (from slide 30) for more information on TF-IDF and Cosine similarity.

## Data Preparation

In [261]:
# (articles_df['title'] + "" + articles_df['text']).iloc[0]

In [262]:
# Ignoring stopwords (words with no semantics) from English and Portuguese (as we have a corpus with mixed languages)
stopwords_list = stopwords.words('english') + stopwords.words('portuguese')

# Trains a model whose vectors size is 5000, composed by the main unigrams and bigrams found in the corpus, ignoring stopwords
vectorizer = TfidfVectorizer(analyzer='word',
                     ngram_range=(1, 2),
                     min_df=0.003,
                     max_df=0.5,
                     max_features=5000,
                     stop_words=stopwords_list)

In [263]:
tfidf_matrix : scipy.sparse.csr.csr_matrix = vectorizer.fit_transform(articles_df['title'] + "" + articles_df['text'])
tfidf_feature_names = vectorizer.get_feature_names()
tfidf_matrix
item_ids: list = articles_df['contentId'].tolist()

In [264]:
len(tfidf_feature_names)

5000

In [265]:
# tfidf_feature_names

In [266]:
tfidf_matrix

<3047x5000 sparse matrix of type '<class 'numpy.float64'>'
	with 638928 stored elements in Compressed Sparse Row format>

### Build Profile for Each User

To model the user profile, we take all the item profiles the user has interacted and average them. The average is weighted by the interaction strength, in other words, the articles the user has interacted the most (eg. liked or commented) will have a higher strength in the final user profile.   

In [267]:
def get_item_profile(item_id):
    idx = item_ids.index(item_id)
    item_profile = tfidf_matrix[idx:idx+1]
    return item_profile

def get_item_profiles(ids):
    item_profiles_list = [get_item_profile(x) for x in ids]
    item_profiles = scipy.sparse.vstack(item_profiles_list)
    return item_profiles

def build_users_profile(person_id, interactions_indexed_df):
    interactions_person_df = interactions_indexed_df.loc[person_id]
    user_item_profiles = get_item_profiles(interactions_person_df['contentId'])
    
    user_item_strengths = np.array(interactions_person_df['eventStrength']).reshape(-1,1)
    # Weighted average of item profiles by the interactions strength
    user_item_strengths_weighted_avg = np.sum(user_item_profiles.multiply(user_item_strengths), axis=0) / np.sum(user_item_strengths)
    user_profile_norm = sklearn.preprocessing.normalize(user_item_strengths_weighted_avg)
    return user_profile_norm

def build_users_profiles():
    msk = ratings_trn_df['contentId'].isin(articles_df['contentId'])
    interactions_indexed_df = ratings_trn_df[msk].set_index('personId')
    user_profiles = {}
    for person_id in interactions_indexed_df.index.unique():
        user_profiles[person_id] = build_users_profile(person_id, interactions_indexed_df)
    return user_profiles

In [268]:
user_profiles = build_users_profiles()

In [269]:
user_profiles[3609194402293569455]

array([[0.01205907, 0.02868777, 0.00373276, ..., 0.00665326, 0.00798772,
        0.00472609]])

In [270]:
# user_profiles[3609194402293569455].flatten().tolist()

In [271]:
len(tfidf_feature_names)

5000

In [272]:
type(tfidf_feature_names)

list

In [273]:
# pd.DataFrame(data = user_profiles, index=tfidf_feature_names)

Let's take a look in the profile. It is a [unit vector](https://en.wikipedia.org/wiki/Unit_vector) of 5000 length. The value in each position represents how relevant is a token (unigram or bigram) for me.  
Looking my profile, it appears that the top relevant tokens really represent my professional interests in **machine learning**, **deep learning**, **artificial intelligence** and **google cloud platform**! So we might expect good recommendations here!

In [274]:
myprofile = user_profiles[-1479311724257856983]

In [275]:
print(myprofile.shape)

(1, 5000)


In [276]:
myprofile_flatten = user_profiles[-1479311724257856983].flatten().tolist()
pd.DataFrame(sorted(zip(tfidf_feature_names, myprofile_flatten), key=lambda x: -x[1]),
    columns=['token', 'relevance'])

token  relevance
0             learning   0.298732
1     machine learning   0.245992
2              machine   0.237843
3               google   0.202839
4                 data   0.169776
...                ...        ...
4995          year old   0.000000
4996         yesterday   0.000000
4997              água   0.000000
4998              área   0.000000
4999             único   0.000000

[5000 rows x 2 columns]

## Develop a Content-Based Recommendation System

In [281]:
class CBRecSys:
    """ Content Based Recommender System. 
    Data Depandency:
        * item_ids     : list
        * tfidf_matrix : scipy.sparse.csr.csr_matrix
        * user_profiles: Dict[int, np.ndarray]    
    """
    
    MODEL_NAME = 'Content-Based'
    
    def __init__(self, items_df = None):
        self.item_ids : list = item_ids
        self.items_df : scipy.sparse.csr.csr_matrix = items_df
        
    def get_model_name(self):
        return self.MODEL_NAME
        
    def _get_similar_items_to_user_profile(self, person_id, topn=1000):
        # Computes the cosine similarity between the user profile and all item profiles
        cosine_similarities = cosine_similarity(user_profiles[person_id], tfidf_matrix)
        
        # Gets the top similar items
        similar_indices = cosine_similarities.argsort().flatten()[-topn:]
        
        # Sort the similar items by similarity
        similar_items = sorted([(item_ids[i], cosine_similarities[0,i]) for i in similar_indices], key=lambda x: -x[1])
        return similar_items
        
    def recommend_items(self, user_id, items_to_ignore=[], topn=10, verbose=False):
        similar_items = self._get_similar_items_to_user_profile(user_id)
        
        # Ignores items the user has already interacted
        similar_items_filtered = list(filter(lambda x: x[0] not in items_to_ignore, similar_items))
        
        recommendations_df = pd.DataFrame(similar_items_filtered, columns=['contentId', 'recStrength']).head(topn)

        if verbose:
            if self.items_df is None: raise Exception('"items_df" is required in verbose mode')

            recommendations_df = recommendations_df.merge(self.items_df, how = 'left', 
                                                          left_on = 'contentId', 
                                                          right_on = 'contentId')[['recStrength', 'contentId', 'title', 'url', 'lang']]


        return recommendations_df
    

In [282]:
content_based_recommender_model = CBRecSys(articles_df)

With personalized recommendations of content-based filtering model, we have a **Recall@5** to about **0.162**, which means that about **16%** of interacted items in test set were ranked by this model among the top-5 items (from lists with 100 random items).
And **Recall@10** was **0.261 (52%)**.
The lower performance of the Content-Based model compared to the Popularity model may indicate that users are not that fixed in content very similar to their previous reads.

In [283]:
print('Evaluating Content-Based Filtering model...')
cb_global_metrics, cb_detailed_results_df = model_evaluator.evaluate_model(content_based_recommender_model)

print('\nGlobal metrics:\n%s' % cb_global_metrics)
cb_detailed_results_df.head(10)

Evaluating Content-Based Filtering model...
1139 users processed

Global metrics:
{'modelName': 'Content-Based', 'recall@5': 0.16287394528253643, 'recall@10': 0.2614420864229097}


hits@5_count  hits@10_count  interacted_count  recall@5  recall@10  \
76             15             24               192  0.078125   0.125000   
17             18             29               134  0.134328   0.216418   
16             20             33               130  0.153846   0.253846   
10             32             47               117  0.273504   0.401709   
82              6             15                88  0.068182   0.170455   
161            11             23                80  0.137500   0.287500   
65              8             13                73  0.109589   0.178082   
81              8             19                69  0.115942   0.275362   
106             3              9                69  0.043478   0.130435   
52              3              8                68  0.044118   0.117647   

              _person_id  
76   3609194402293569455  
17  -2626634673110551643  
16  -1032019229384696495  
10  -1443636648652872475  
82  -2979881261169775358  
161 -3596626804281480007  
65   1116121227607581999  
81    692689608292948411  
106 -9016528795238256703  
52   3636910968448833585

# Collaborative Filtering model

Collaborative Filtering (CF) has two main implementation strategies:  
- **Memory-based**: This approach uses the memory of previous users interactions to compute users similarities based on items they've interacted (<font color='red'>user-based approach</font>) or compute items similarities based on the users that have interacted with them (item-based approach).  
A typical example of this approach is User Neighbourhood-based CF, in which the top-N similar users (usually computed using Pearson correlation) for a user are selected and used to recommend items those similar users liked, but the current user have not interacted yet. This approach is very simple to implement, but usually do not scale well for many users. A nice Python implementation of this approach in available in [Crab](http://muricoca.github.io/crab/).
- **Model-based**: This approach, models are developed using different machine learning algorithms to recommend items to users. There are many model-based CF algorithms, like neural networks, bayesian networks, clustering models, and latent factor models such as Singular Value Decomposition (SVD) and, probabilistic latent semantic analysis.

## Matrix Factorization

- [Lecture 54 — Latent Factor Models | Stanford University](https://www.youtube.com/watch?v=4-f77HjB_CI&list=PLLssT5z_DsK9JDLcT8T62VtzwyW9LNepV&index=54)
- [Lecture 55 — Latent Factor Recommender System | Stanford University](https://www.youtube.com/watch?v=E8aMcwmqsTg&list=PLLssT5z_DsK9JDLcT8T62VtzwyW9LNepV&index=55)

Latent factor models compress user-item matrix into a low-dimensional representation in terms of latent factors. One advantage of using this approach is that instead of having a high dimensional matrix containing abundant number of missing values we will be dealing with a much smaller matrix in lower-dimensional space.  
A reduced presentation could be utilized for either user-based or item-based neighborhood algorithms that are presented in the previous section. There are several advantages with this paradigm. It handles the sparsity of the original matrix better than memory based ones. Also comparing similarity on the resulting matrix is much more scalable especially in dealing with large sparse datasets.  

Here we a use popular latent factor model named [Singular Value Decomposition (SVD)](https://en.wikipedia.org/wiki/Singular_value_decomposition). There are other matrix factorization frameworks more specific to CF you might try, like [surprise](https://github.com/NicolasHug/Surprise), [mrec](https://github.com/Mendeley/mrec) or [python-recsys](https://github.com/ocelma/python-recsys). We chose a [SciPy](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.svds.html) implemenation of SVD because it is available on Kaggle kernels. 
P.s. See an example of SVD on a movies dataset in this [blog post](https://beckernick.github.io/matrix-factorization-recommender/).   

An important decision is the number of factors to factor the user-item matrix. The higher the number of factors, the more precise is the factorization in the original matrix reconstructions. Therefore, if the model is allowed to  memorize too much details of the original matrix, it may not generalize well for data it was not trained on. Reducing the number of factors increases the model generalization.

In [285]:
# Creating a sparse pivot table with users in rows and items in columns
users_items_pivot_matrix_df = ratings_trn_df.pivot(index='personId', 
                                                          columns='contentId', 
                                                          values='eventStrength').fillna(0)

In [286]:
users_items_pivot_matrix_df.head(10)

contentId             -9222795471790223670  -9216926795620865886  \
personId                                                           
-9223121837663643404                   0.0                   0.0   
-9212075797126931087                   0.0                   0.0   
-9207251133131336884                   0.0                   2.0   
-9199575329909162940                   0.0                   0.0   
-9196668942822132778                   0.0                   0.0   
-9188188261933657343                   0.0                   0.0   
-9172914609055320039                   0.0                   0.0   
-9156344805277471150                   0.0                   0.0   
-9120685872592674274                   0.0                   0.0   
-9109785559521267180                   0.0                   0.0   

contentId             -9194572880052200111  -9192549002213406534  \
personId                                                           
-9223121837663643404                   0.0                   0.0   
-9212075797126931087                   0.0                   0.0   
-9207251133131336884                   0.0                   0.0   
-9199575329909162940                   0.0                   0.0   
-9196668942822132778                   0.0                   0.0   
-9188188261933657343                   0.0                   0.0   
-9172914609055320039                   0.0                   0.0   
-9156344805277471150                   0.0                   0.0   
-9120685872592674274                   0.0                   0.0   
-9109785559521267180                   0.0                   0.0   

contentId             -9190737901804729417  -9189659052158407108  \
personId                                                           
-9223121837663643404                   0.0                   0.0   
-9212075797126931087                   0.0                   0.0   
-9207251133131336884                   0.0                   0.0   
-9199575329909162940                   0.0                   0.0   
-9196668942822132778                   0.0                   0.0   
-9188188261933657343                   0.0                   0.0   
-9172914609055320039                   0.0                   0.0   
-9156344805277471150                   0.0                   0.0   
-9120685872592674274                   0.0                   0.0   
-9109785559521267180                   0.0                   0.0   

contentId             -9176143510534135851  -9172673334835262304  \
personId                                                           
-9223121837663643404                   0.0                   0.0   
-9212075797126931087                   0.0                   0.0   
-9207251133131336884                   0.0                   0.0   
-9199575329909162940                   0.0                   0.0   
-9196668942822132778                   0.0                   0.0   
-9188188261933657343                   0.0                   0.0   
-9172914609055320039                   0.0                   0.0   
-9156344805277471150                   0.0                   0.0   
-9120685872592674274                   0.0                   0.0   
-9109785559521267180                   0.0                   0.0   

contentId             -9171475473795142532  -9166778629773133902  ...  \
personId                                                          ...   
-9223121837663643404                   0.0                   0.0  ...   
-9212075797126931087                   0.0                   0.0  ...   
-9207251133131336884                   0.0                   0.0  ...   
-9199575329909162940                   0.0                   0.0  ...   
-9196668942822132778                   0.0                   0.0  ...   
-9188188261933657343                   0.0                   0.0  ...   
-9172914609055320039                   0.0                   0.0  ...   
-9156344805277471150                   0.0                   0.0  ...   
-9

In [287]:
users_items_pivot_matrix_df.shape

(1140, 2926)

In [288]:
users_items_pivot_matrix = users_items_pivot_matrix_df.values
users_items_pivot_matrix[:10]

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 2., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])

In [289]:
users_ids = list(users_items_pivot_matrix_df.index)
users_ids[:10]

[-9223121837663643404,
 -9212075797126931087,
 -9207251133131336884,
 -9199575329909162940,
 -9196668942822132778,
 -9188188261933657343,
 -9172914609055320039,
 -9156344805277471150,
 -9120685872592674274,
 -9109785559521267180]

In [290]:
users_items_pivot_sparse_matrix = csr_matrix(users_items_pivot_matrix)
users_items_pivot_sparse_matrix

<1140x2926 sparse matrix of type '<class 'numpy.float64'>'
	with 31284 stored elements in Compressed Sparse Row format>

In [291]:
# The number of factors to factor the user-item matrix.
NUMBER_OF_FACTORS_MF = 15
# Performs matrix factorization of the original user item matrix
# U, sigma, Vt = svds(users_items_pivot_matrix, k = NUMBER_OF_FACTORS_MF)
U, sigma, Vt = svds(users_items_pivot_sparse_matrix, k = NUMBER_OF_FACTORS_MF)

In [292]:
U.shape

(1140, 15)

In [293]:
Vt.shape

(15, 2926)

In [294]:
sigma = np.diag(sigma)
sigma.shape

(15, 15)

After the factorization, we try to to reconstruct the original matrix by multiplying its factors. The resulting matrix is not sparse any more. It was generated predictions for items the user have not yet interaction, which we will exploit for recommendations.

In [295]:
all_user_predicted_ratings = np.dot(np.dot(U, sigma), Vt) 
all_user_predicted_ratings

array([[ 0.01039915,  0.00081872, -0.01725263, ...,  0.00140708,
         0.0110647 ,  0.00226063],
       [-0.00019285, -0.00031318, -0.00264624, ...,  0.00251658,
         0.00017609, -0.00189488],
       [-0.01254721,  0.0065947 , -0.00590676, ...,  0.00698975,
        -0.01015696,  0.01154572],
       ...,
       [-0.02995379,  0.00805715, -0.01846307, ..., -0.01083078,
        -0.00118591,  0.0096798 ],
       [-0.01845505,  0.00467019,  0.01219602, ...,  0.00409507,
         0.00019482, -0.00752562],
       [-0.01506374,  0.00327732,  0.13391269, ..., -0.01191815,
         0.06422074,  0.01303244]])

In [296]:
all_user_predicted_ratings_norm = (all_user_predicted_ratings - all_user_predicted_ratings.min()) / (all_user_predicted_ratings.max() - all_user_predicted_ratings.min())

In [297]:
# Converting the reconstructed matrix back to a Pandas dataframe
cf_preds_df = pd.DataFrame(all_user_predicted_ratings_norm, columns = users_items_pivot_matrix_df.columns, index=users_ids).transpose()
cf_preds_df.head(10)

-9223121837663643404  -9212075797126931087  \
contentId                                                          
-9222795471790223670              0.139129              0.137930   
-9216926795620865886              0.138044              0.137916   
-9194572880052200111              0.135998              0.137652   
-9192549002213406534              0.141924              0.137996   
-9190737901804729417              0.140209              0.137408   
-9189659052158407108              0.138932              0.138699   
-9176143510534135851              0.143208              0.138673   
-9172673334835262304              0.138527              0.138021   
-9171475473795142532              0.140720              0.137865   
-9166778629773133902              0.138989              0.137725   

                      -9207251133131336884  -9199575329909162940  \
contentId                                                          
-9222795471790223670              0.136531              0.143948   
-9216926795620865886              0.138698              0.137878   
-9194572880052200111              0.137283              0.137536   
-9192549002213406534              0.134663              0.137080   
-9190737901804729417              0.138708              0.138672   
-9189659052158407108              0.138117              0.137621   
-9176143510534135851              0.139514              0.139114   
-9172673334835262304              0.138274              0.137827   
-9171475473795142532              0.138061              0.137633   
-9166778629773133902              0.136520              0.137723   

                      -9196668942822132778  -9188188261933657343  \
contentId                                                          
-9222795471790223670              0.136815              0.137339   
-9216926795620865886              0.137969              0.137990   
-9194572880052200111              0.140363              0.137807   
-9192549002213406534              0.139946              0.138574   
-9190737901804729417              0.137725              0.138218   
-9189659052158407108              0.138920              0.137766   
-9176143510534135851              0.137664              0.137447   
-9172673334835262304              0.137997              0.138037   
-9171475473795142532              0.138231              0.138089   
-9166778629773133902              0.138559              0.137951   

                      -9172914609055320039  -9156344805277471150  \
contentId                                                          
-9222795471790223670              0.137508              0.143534   
-9216926795620865886              0.137974              0.138049   
-9194572880052200111              0.141246              0.136284   
-9192549002213406534              0.139473              0.144469   
-9190737901804729417              0.138390              0.138418   
-9189659052158407108              0.138568              0.138200   
-9176143510534135851              0.139833              0.140564   
-9172673334835262304              0.138104              0.138259   
-9171475473795142532              0.139009              0.137552   
-9166778629773133902              0.138189              0.138496   

                      -9120685872592674274  -9109785559521267180  ...  \
contentId                                                         ...   
-9222795471790223670              0.136428              0.135681  ...   
-9216926795620865886              0.138217              0.138151  ...   
-9194572880052200111              0.135301              0.138512  ...   
-9192549002213406534              0.143333              0.138428  ...   
-9190737901804729417              0.134883              0.140193  ...   
-9189659052158407108              0.140572              0.140365  ...   
-9176143510534135851              0.144698              0.144440  ...   
-9172673334835262304              0.137633              0.138397  ...   
-9171475473795142532    

In [298]:
cf_preds_df.shape

(2926, 1140)

In [299]:
class CFRecSys:
    
    MODEL_NAME = 'Collaborative Filtering'
    
    def __init__(self, cf_predictions_df, items_df=None):
        self.cf_predictions_df = cf_predictions_df
        self.items_df = items_df
        
    def get_model_name(self):
        return self.MODEL_NAME
        
    def recommend_items(self, user_id, items_to_ignore=[], topn=10, verbose=False):
        # Get and sort the user's predictions
        sorted_user_predictions = self.cf_predictions_df[user_id].sort_values(ascending=False) \
                                    .reset_index().rename(columns={user_id: 'recStrength'})

        # Recommend the highest predicted rating movies that the user hasn't seen yet.
        recommendations_df = sorted_user_predictions[~sorted_user_predictions['contentId'].isin(items_to_ignore)] \
                               .sort_values('recStrength', ascending = False) \
                               .head(topn)

        if verbose:
            if self.items_df is None:
                raise Exception('"items_df" is required in verbose mode')

            recommendations_df = recommendations_df.merge(self.items_df, how = 'left', 
                                                          left_on = 'contentId', 
                                                          right_on = 'contentId')[['recStrength', 'contentId', 'title', 'url', 'lang']]


        return recommendations_df
    

In [300]:
cf_recommender_model = CFRecSys(cf_preds_df, articles_df)

> Evaluating the Collaborative Filtering model (SVD matrix factorization), we observe that we got **Recall@5 (33%)** and **Recall@10 (46%)** values, much higher than Popularity model and Content-Based model.

In [301]:
print('Evaluating Collaborative Filtering (SVD Matrix Factorization) model...')
cf_global_metrics, cf_detailed_results_df = model_evaluator.evaluate_model(cf_recommender_model)
print('\nGlobal metrics:\n%s' % cf_global_metrics)
cf_detailed_results_df.head(10)

Evaluating Collaborative Filtering (SVD Matrix Factorization) model...
1139 users processed

Global metrics:
{'modelName': 'Collaborative Filtering', 'recall@5': 0.33392994119151115, 'recall@10': 0.46803886474047557}


hits@5_count  hits@10_count  interacted_count  recall@5  recall@10  \
76             21             46               192  0.109375   0.239583   
17             30             56               134  0.223881   0.417910   
16             16             34               130  0.123077   0.261538   
10             38             51               117  0.324786   0.435897   
82             39             48                88  0.443182   0.545455   
161            22             34                80  0.275000   0.425000   
65             24             32                73  0.328767   0.438356   
81             16             21                69  0.231884   0.304348   
106            20             28                69  0.289855   0.405797   
52             23             30                68  0.338235   0.441176   

              _person_id  
76   3609194402293569455  
17  -2626634673110551643  
16  -1032019229384696495  
10  -1443636648652872475  
82  -2979881261169775358  
161 -3596626804281480007  
65   1116121227607581999  
81    692689608292948411  
106 -9016528795238256703  
52   3636910968448833585

## Hybrid Recommender

What if we combine Collaborative Filtering and Content-Based Filtering approaches?    
Would that provide us with more accurate recommendations?    
In fact, hybrid methods have performed better than individual approaches in many studies and have being extensively used by researchers and practioners.  
Let's build a simple hybridization method, as an ensemble that takes the weighted average of the normalized CF scores with the Content-Based scores, and ranking by resulting score. In this case, as the CF model is much more accurate than the CB model, the weights for the CF and CB models are 100.0 and 1.0, respectivelly.

In [ ]:
class HybridRecommender:
    
    MODEL_NAME = 'Hybrid'
    
    def __init__(self, cb_rec_model, cf_rec_model, items_df, cb_ensemble_weight=1.0, cf_ensemble_weight=1.0):
        self.cb_rec_model = cb_rec_model
        self.cf_rec_model = cf_rec_model
        self.cb_ensemble_weight = cb_ensemble_weight
        self.cf_ensemble_weight = cf_ensemble_weight
        self.items_df = items_df
        
    def get_model_name(self):
        return self.MODEL_NAME
        
    def recommend_items(self, user_id, items_to_ignore=[], topn=10, verbose=False):
        #Getting the top-1000 Content-based filtering recommendations
        cb_recs_df = self.cb_rec_model.recommend_items(user_id, items_to_ignore=items_to_ignore, verbose=verbose,
                                                           topn=1000).rename(columns={'recStrength': 'recStrengthCB'})
        
        #Getting the top-1000 Collaborative filtering recommendations
        cf_recs_df = self.cf_rec_model.recommend_items(user_id, items_to_ignore=items_to_ignore, verbose=verbose, 
                                                           topn=1000).rename(columns={'recStrength': 'recStrengthCF'})
        
        #Combining the results by contentId
        recs_df = cb_recs_df.merge(cf_recs_df,
                                   how = 'outer', 
                                   left_on = 'contentId', 
                                   right_on = 'contentId').fillna(0.0)
        
        #Computing a hybrid recommendation score based on CF and CB scores
        #recs_df['recStrengthHybrid'] = recs_df['recStrengthCB'] * recs_df['recStrengthCF'] 
        recs_df['recStrengthHybrid'] = (recs_df['recStrengthCB'] * self.cb_ensemble_weight) \
                                     + (recs_df['recStrengthCF'] * self.cf_ensemble_weight)
        
        #Sorting recommendations by hybrid score
        recommendations_df = recs_df.sort_values('recStrengthHybrid', ascending=False).head(topn)

        if verbose:
            if self.items_df is None:
                raise Exception('"items_df" is required in verbose mode')

            recommendations_df = recommendations_df.merge(self.items_df, how = 'left', 
                                                          left_on = 'contentId', 
                                                          right_on = 'contentId')[['recStrengthHybrid', 'contentId', 'title', 'url', 'lang']]


        return recommendations_df

In [ ]:
hybrid_recommender_model = HybridRecommender(content_based_recommender_model, cf_recommender_model, articles_df,
                                             cb_ensemble_weight=1.0, cf_ensemble_weight=100.0)

**We have a new champion!**  
Our simple hybrid approach surpasses Content-Based filtering with its combination with Collaborative Filtering. Now we have a **Recall@5** of **34.2%** and **Recall@10** of **47.9%**

In [ ]:
print('Evaluating Hybrid model...')
hybrid_global_metrics, hybrid_detailed_results_df = model_evaluator.evaluate_model(hybrid_recommender_model)
print('\nGlobal metrics:\n%s' % hybrid_global_metrics)
hybrid_detailed_results_df.head(10)

## Comparing the methods

In [ ]:
global_metrics_df = pd.DataFrame([cb_global_metrics, pop_global_metrics, cf_global_metrics, hybrid_global_metrics]) \
                        .set_index('modelName')
global_metrics_df

In [ ]:
%matplotlib inline
ax = global_metrics_df.transpose().plot(kind='bar', figsize=(15,8))
for p in ax.patches:
    ax.annotate("%.3f" % p.get_height(), (p.get_x() + p.get_width() / 2., p.get_height()), ha='center', va='center', xytext=(0, 10), textcoords='offset points')

# Testing

Let's test the best model (Hybrid) for my user.

In [ ]:
def inspect_interactions(person_id, test_set=True):
    if test_set:
        interactions_df = interactions_test_indexed_df
    else:
        interactions_df = interactions_train_indexed_df
    return interactions_df.loc[person_id].merge(articles_df, how = 'left', 
                                                      left_on = 'contentId', 
                                                      right_on = 'contentId') \
                          .sort_values('eventStrength', ascending = False)[['eventStrength', 
                                                                          'contentId',
                                                                          'title', 'url', 'lang']]

Here we see some articles I interacted in Deskdrop from train set. It can be easily observed that among my main interests are **machine learning**, **deep learning**, **artificial intelligence**, and **google cloud platform**.

In [ ]:
inspect_interactions(-1479311724257856983, test_set=False).head(20)

**The recommendations really matches my interests, as I would read all of them!**

In [ ]:
hybrid_recommender_model.recommend_items(-1479311724257856983, topn=20, verbose=True)

# Conclusion

In this notebook, we've explored and compared the main Recommender Systems techniques on [CI&T Deskdrop](https://www.kaggle.com/gspmoreira/articles-sharing-reading-from-cit-deskdrop) dataset. It could be observed that for articles recommendation, content-based filtering and a hybrid method performed better than Collaborative Filtering alone.  

There is large room for improvements of the results. Here are some tips:
- In this example, we've completely ignored the time, considering that all articles were available to be recommended to users at any time. A better approach would be to filter only articles that were available for users at a given time.
- You could leverage the available contextual information to model users preferences across time (period of day, day of week, month), location (country and state/district) and devices (browser, mobile native app).  
This contextual information can be easily incorporated in [Learn-to-Rank](https://en.wikipedia.org/wiki/Learning_to_rank) models (like XGBoost Gradient Boosting Decision Trees with ranking objective), Logistic models (with categorical features [One-Hot encoded](http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.OneHotEncoder.html) or [Feature Hashed](https://en.wikipedia.org/wiki/Feature_hashing)), and [Wide & Deep models](https://ai.googleblog.com/2016/06/wide-deep-learning-better-together-with.html), which is implemented in [TensorFlow](https://docs.w3cub.com/tensorflow~guide/tutorials/wide_and_deep/). Take a look in the summary my solution shared for [Outbrain Click Prediction](https://www.kaggle.com/c/outbrain-click-prediction/discussion/27897#157215) competition. 
- Those basic techniques were used for didactic purposes. There are more advanced techniques in RecSys research community, specially advanced Matrix Factorization and Deep Learning models.  

You can know more about state-of-the-art methods published in Recommender Systems on [ACM RecSys conference](https://recsys.acm.org/).  
If you are more like practioner than researcher, you might try some Collaborative Filtering frameworks in this dataset, like [surprise](https://github.com/NicolasHug/Surprise), [mrec](https://github.com/Mendeley/mrec),  [python-recsys](https://github.com/ocelma/python-recsys) and [Spark ALS Matrix Factorization](https://spark.apache.org/docs/latest/mllib-collaborative-filtering.html) (distributed implementation for large datasets).  
Take a look in this [presentation](https://www.slideshare.net/gabrielspmoreira/discovering-users-topics-of-interest-in-recommender-systems-tdc-sp-2016) where I describe a production recommender system, focused on Content-Based Filtering and Topic Modeling techniques.